hpx/parallel/util/numa_binding_allocator.hpp

```diff
@@ -0,0 +1,569 @@
+
+        // The number of elements along dimension x=0,y=1,z=2,...
+        // This function is only required for debug purposes
+        virtual std::size_t array_size(std::size_t axis) const {
+            return memory_bytes()/sizeof(T);
+        };
+
+        // When counting along elements in a given dimension,
+        // how large a step should be taken in units of elements.
+        // This should include padding along an axis
+        // This function is only required for debug purposes
+        virtual std::size_t memory_step(std::size_t axis) const { return 1; };
+
+        // When displaying the data, what step size should be used
+        // This function is only required for debug purposes
+        virtual std::size_t display_step(std::size_t axis) const { return 1; };
+
+        // This is for debug/information purposes only and anything may
+        // be returned that helps illuminate the mapping of pages->indices
+        virtual std::string description() const { return ""; };
+//#endif
+
+        std::string pool_name_ = "default";
+    };
+
+    template <typename T>
+    using numa_binding_helper_ptr = std::shared_ptr<numa_binding_helper<T>>;
+
+    /// The numa_binding_allocator allocates memory using a policy based on
+    /// hwloc flags for memory binding.
+    /// This allocator can be used to request data that is bound
+    /// to one or more numa domains via the bitmap mask supplied
+
+    template <typename T >
+    struct numa_binding_allocator
+    {
+        typedef T              value_type;
+        typedef T*             pointer;
+        typedef const T*       const_pointer;
+        typedef T&             reference;
+        typedef T const&       const_reference;
+        typedef std::size_t    size_type;
+        typedef std::ptrdiff_t difference_type;
+        //
+        template <typename U>
+        struct rebind
+        {
+            typedef numa_binding_allocator<U> other;
+        };
+
+        using numa_binding_helper_ptr = std::shared_ptr<numa_binding_helper<T>>;
+
+        // default constructor produces an unusable allocator
+        numa_binding_allocator() {}
+
+        // construct without a memory binder function
+        // only first touch or interleave policies are valid
+        numa_binding_allocator(threads::hpx_hwloc_membind_policy policy,
+                       unsigned int flags)
+            : policy_(policy)
+            , flags_(flags)
+        {
+            LOG_NUMA_MSG("numa_binding_allocator : no binder function");
+            HPX_ASSERT(policy_ != threads::hpx_hwloc_membind_policy::membind_user);
+        }
+
+        // construct using a memory binder function for placing pages
+        // according to a user defined pattern using first touch policy
+        numa_binding_allocator(numa_binding_helper_ptr bind_func,
+                       threads::hpx_hwloc_membind_policy policy,
+                       unsigned int flags)
+            : binding_helper_(bind_func)
```

Comment:
This could use `std::move()` to avoid the (atomic) reference count.

Reply:
you mean the shared_ptr? I would not worry here. this is not a critical piece of code, the allocator will be constructed once before doing math operations and kept alive until termination most likely